crates/primitives/src/env.rs

```diff
@@ -128,65 +140,66 @@ impl Env {
-                }
+            // The field `to` deviates slightly from the semantics with the exception
+            // that it MUST NOT be nil and therefore must always represent
+            // a 20-byte address. This means that blob transactions cannot
+            // have the form of a create transaction.
+            if self.tx.transact_to.is_create() {
+                return Err(InvalidTransaction::BlobCreateTransaction);
+            }
 
-                // all versioned blob hashes must start with VERSIONED_HASH_VERSION_KZG
-                for blob in self.tx.blob_hashes.iter() {
-                    if blob[0] != VERSIONED_HASH_VERSION_KZG {
-                        return Err(InvalidTransaction::BlobVersionNotSupported);
-                    }
+            // all versioned blob hashes must start with VERSIONED_HASH_VERSION_KZG
+            for blob in self.tx.blob_hashes.iter() {
+                if blob[0] != VERSIONED_HASH_VERSION_KZG {
+                    return Err(InvalidTransaction::BlobVersionNotSupported);
                 }
+            }
 
-                // ensure the total blob gas spent is at most equal to the limit
-                // assert blob_gas_used <= MAX_BLOB_GAS_PER_BLOCK
-                let num_blobs = self.tx.blob_hashes.len();
-                if num_blobs > MAX_BLOB_NUMBER_PER_BLOCK as usize {
-                    return Err(InvalidTransaction::TooManyBlobs {
-                        have: num_blobs,
-                        max: MAX_BLOB_NUMBER_PER_BLOCK as usize,
-                    });
-                }
+            // ensure the total blob gas spent is at most equal to the limit
+            // assert blob_gas_used <= MAX_BLOB_GAS_PER_BLOCK
+            let num_blobs = self.tx.blob_hashes.len();
+            if num_blobs > MAX_BLOB_NUMBER_PER_BLOCK as usize {
+                return Err(InvalidTransaction::TooManyBlobs {
+                    have: num_blobs,
+                    max: MAX_BLOB_NUMBER_PER_BLOCK as usize,
+                });
             }
         } else {
+            // if max_fee_per_blob_gas is not set, then blob_hashes must be empty
             if !self.tx.blob_hashes.is_empty() {
                 return Err(InvalidTransaction::BlobVersionedHashesNotSupported);
             }
-            if self.tx.max_fee_per_blob_gas.is_some() {
-                return Err(InvalidTransaction::MaxFeePerBlobGasNotSupported);
+        }
+
+        // check if EIP-7702 transaction is enabled.
+        if !SPEC::enabled(SpecId::PRAGUE) && self.tx.authorization_list.is_some() {
+            return Err(InvalidTransaction::AuthorizationListNotSupported);
+        }
+
+        if self.tx.authorization_list.is_some() {
+            // Check if other fields are unset.
+            if self.tx.max_fee_per_blob_gas.is_some() || !self.tx.blob_hashes.is_empty() {
+                return Err(InvalidTransaction::AuthorizationListInvalidFields);
```

Comment:
should this not be `BlobVersionedHashesNotSupported`? or something similar, this error is a bit vague since the error is that we have an auth list *and* blobs

Reply:
Yeah, EnvTx sucks rn, as we have all fields in one struck. This error is here if it is AuthorizationList TX don't set other fields by mistake.